Clean up and simplify camera override API

rxlecky opened the pull request:
- Make the camera override 2D and 3D APIs use the same terminology
- Make both 2D and 3D override use Camera nodes under the hood to make the code simpler and cleaner
- Make camera override 2D use global canvas transform to make the override camera behave identically to the 2D editor camera, i.e. make CanvasLayers move with the camera regardless of their follow_viewport property value
- Makes all code depending on the viewport current camera, i.e. VisibleOnScreenNotifiers, work with the camera override out of the box

The change to using camera nodes also has potential downsides. If there's gameplay code depending on current camera, it might cause unexpected behaviour when the current camera suddenly changes to the overriden camera. I had a person ask me whether using the camera override will make the environment unload if they are using visibility culling, causing, for example, the player character to fall through the floor. Similar arguments were the reason, why I originally opted for only creating camera on the RenderingServer without creating the actual camera node. That way, everything would work as if the override camera never existed. 

However, reevaluating that now, duplicate code had to be written to support the same functions that the camera nodes already support. Additionally, there are cases where we do want the override camera to work as normal camera, example would be the case described in the issue #49334. All these would have to be catered to separately if the camera override didn't use the camera nodes, whereas now they all "just work". Lastly, to follow up on the visibility culling argument, I think the camera override wouldn't be really that useful if nothing besides the area in player's close vicinity would be loaded and rendered anyway.

Any feedback is welcome. If there's another approach that I haven't thought of, I'm happy to rework this. Cheers!

Closes #49334

rxlecky commented:
Oh, actually, #49334 won't be fixed since that's a 3.3.2 bug. Happy to backport this once it's approved and merged, though.

akien-mga commented on scene/main/viewport.h:
```diff
@@ -204,16 +204,31 @@ class Viewport : public Node {
+#ifdef TOOLS_ENABLED
+	struct Camera2DOverride {
+		bool enabled = false;
+		ObjectID previous_camera = ObjectID();
```
`ObjectID` is default initialized so you don't need to assign a new one here (same below).

rxlecky replied:
Ah, good spot! Thanks, fixed now. 🙂

rxlecky commented:
@akien-mga unless there are further comments on code cleanliness, this should be good to be merged. But I'd still wait before merging. I'm not entirely sure whether it was @reduz 's suggestion back when this was first implemented to not use nodes or it was my initiative. It would be good to get his - or other experienced member's - opinion on whether this is a good and desired change before merging.

rxlecky commented:
@akien-mga do you think this could get reviewed sometime soon? I'm happy to merge and rebase but I'm putting it off for now until someone has the time for review.

KoBeWi commented:
Camera overrides were replaced by the Game tab: #97257
The linked issue still needs fixing though.

rxlecky commented:
Alright, I got around to doing this quite a bit later than I wanted but I think I've got the changes ready now. I started from scractch but the original goals stayed mostly unchanged, including fixing the linked issue.

The only goal that's changed is switching to using the global transform for override 2D. The change made sense at the time when the only way to control the override camera was via the 2D editor camera which uses global transform. The difference in behaviours made for a bit jarring user experience. But now that the override camera can be controlled directly from the Game tab, I think leaving it as is is the better option to leave it consistent with regular cameras. My understanding is that the editor camera style of control is only left for compatibility and/or systems that don't support window embedding anyway.

rxlecky commented:
Now, let's see if I can wrangle git to reset the branch to be up to date with master first. 😆

rxlecky committed:
Cleanup and simplify camera override API
- Harmonise the camera override 2D and 3D APIs
- Switch to using Camera2D/3D nodes to provide override functionality. This makes for simpler code, gets rid of much of copy-pasted camera code and makes code that relies on current viewport camera such as VisibleOnScreenNotifiers and object picking work out of the box.
- Make camera override code only accesible within DEBUG_ENABLED builds

KoBeWi commented:
2D zooming is broken. It does not zoom on cursor.

https://github.com/user-attachments/assets/5d0b0f96-bc9b-4fc9-af01-beefa0df1476

Expected (in editor):

https://github.com/user-attachments/assets/00ce6bae-ba34-4956-ae60-f96bcbb6f719

rxlecky commented:
Well spotted, thanks for testing! I'll have a look at fixing this asap.

rxlecky commented:
Alright, 2D zooming should be fixed now.

KoBeWi reviewed (approved):
Works correctly, though I tested mainly 2D.
I checked the code and looks like nice simplification.

Calinou commented:
> I had a person ask me whether using the camera override will make the environment unload if they are using visibility culling, causing, for example, the player character to fall through the floor. 

This particular scenario can be handled in your character controller by freezing the character if they stand in an unloaded area. The floor will reappear below their feet when the camera gets close enough again to reload the area.

Calinou reviewed (commented):
Tested locally, it works as expected.

Example in the [3D Waypoints](https://github.com/godotengine/godot-demo-projects/tree/master/3d/waypoints) demo, which uses projected Controls in 2D space to match the 3D world:

### Before

https://github.com/user-attachments/assets/15adb420-12c4-4629-ba6d-694798a11b56

### After

https://github.com/user-attachments/assets/096ab371-1633-4f2c-b4f9-3292236de982

However, when I uncheck the camera override after moving the camera and switching to Input mode, I will always get a script error:

![image](https://github.com/user-attachments/assets/4111e2b1-f4e6-49ce-b29b-1be86f0c9911)

Video that shows how to reproduce the error (it's most reproducible if you toggle the camera override on and off quickly):

https://github.com/user-attachments/assets/b0012bae-5ddd-44d3-962d-933a4a6c1c55

Adding a `null` check in the project code avoids it, but not every project will have this check in place. Also, if I add a `null` check to early return at the top of `_process()`, the whole block will eventually stop running when I would have gotten an error without the `null` check:

https://github.com/user-attachments/assets/1ffbe660-a118-4826-a4c9-bb3015fb916d

rxlecky commented:
> > I had a person ask me whether using the camera override will make the environment unload if they are using visibility culling, causing, for example, the player character to fall through the floor.
> 
> This particular scenario can be handled in the character controller by freezing the character if they stand in an unloaded area. The floor will reappear below their feet when the camera gets close enough again to reload the area.

I'm sure there's a way to work around this. It just feels wrong to me that in order to use an editor tool the user needs to add code to handle that special case. 🤔 I'm not saying we shouldn't follow through with the changes for now but it may be worth looking into possible out-of-the-box solutions. For example, investigating a viability of adding a feature in visibility notifiers/enablers that would allow choosing whether they should only consider the override camera or both the override and overridden cameras for determining visibility.

Also, I intentionally didn't expose the override API in the scripting bindings because, again, I think the usage of the tool should ideally be transparent to the user and I find it to be quite an elegant solution that `Viewport` reports the override camera as it would report any other current camera. But if we expect users having to handle the override camera as a special case, maybe it would be worth exposing the override API to give users the tools they need to handle it?

rxlecky commented:
> when I uncheck the camera override after moving the camera and switching to Input mode, I will always get a script error

The error is thrown because the camera instance is freed when override is turned off and a new one is created when it's turned on again. I find it more straightforward to just start with a fresh instance rather than keeping the old one around to be reused later. I can change it if caching instances is the more common approach in the codebase or more desirable in this situation. But I would argue that in this case the problem is that the script is caching and using an instance that gives no guarantees about its lifetime. The correct approach would be to either verify it's still valid using `is_instance_valid(...)` before use or query `Viewport` for the current camera every time it's needed.

rxlecky commented:
Also a side-effect of this change is that the override cameras now show up in the remote scene tree when camera override is in use:
![image](https://github.com/user-attachments/assets/7e96301e-ca3f-4044-8172-f6e1db366c4f)

... and frankly, I'm not the biggest fan of this. It's an unnecessary visual clutter in the best case and a way for user to unwittingly change things and break the tool or crash the editor in the worst. It made me wonder if internal child nodes should not be hidden in the remote tree? Or would that cause an issue if such node ever needs to be displayed in the inspector?

rxlecky commented:
> Remote tree is supposed to show all nodes.

If it was a deliberate decision then I have nothing to add! 🙂

> You can't delete remote nodes and changing it won't crash anything.

Alright, if it's not possible to delete the nodes then there shouldn't be a way to cause a crash, you're right. The worst that can happen is changing some of the camera properties could mess up the camera controls and behaviour. But since the camera nodes are not cached and always recreated as mentioned above, the benefit is that it's easy to reset the system and get it back to a working state by just turning the override off and on again.

KoBeWi commented:
I think the issue Calinou mentioned is acceptable, but maybe it's worth to mention it in the description of `get_camera_3d()`.

rxlecky commented:
Good point! I updated the docs now - I welcome any feedback. 🙂

AThousandShips commented on scene/main/viewport.cpp:
```diff
@@ -5594,3 +5484,64 @@ void SubViewport::_validate_property(PropertyInfo &p_property) const {
+// Explicit template instantiation to allow template definitions inside cpp file
+// and prevent instantiation using other than the desired camera types.
+template class Viewport::CameraOverride<Camera2D>;
+template class Viewport::CameraOverride<Camera3D>;
```
That PR doesn't fix this currently, it still needs work so https://github.com/godotengine/godot/pull/111149 is still a proper fix for this
